qmi/instruments/picotech/picoscope3403.py

```diff
@@ -177,46 +215,49 @@ def acquire_by_trigger_and_get_data(
 
             self.set_channel(chan, True, couplings[chan], ranges, 0)
 
-        number_samples = int(time_span / time_base)
+        # Calculate the total number of samples in the time span.
+        number_samples = time_span // sampling_interval
+        _logger.info(f"Timespan = %i Num samples = %i", time_span, number_samples)
+
         if trigger_channel is not None:
             self.stop()
             trig_int = int(self.MAX_SAMPLE_VALUE / max_volts[channels.index(trigger_channel)] * trigger_level)
             self.set_trigger(True, trigger_channel, trig_int, TriggerEdge.RISING)
-            self.run_block(number_samples, number_samples, time_base + 2)
+            self.run_block(number_samples // 2, number_samples - number_samples // 2, time_base)
 
         else:
             self.disable_trigger()
             self.stop()
-            self.run_block(number_samples * 2, 0, time_base + 2)
+            self.run_block(number_samples, 0, time_base)
 
         self.wait_block_ready(timeout=1.5)
         data = self.get_block_data(channels)
-        timetrace = np.arange(0, 2 * number_samples, 1)
+        times = np.arange(0, number_samples, 1) * sampling_interval
         voltages = list()
         for i in channels:
             voltages.append(data[0][channels.index(i)] * (max_volts[channels.index(i)] / self.MAX_SAMPLE_VALUE))
 
-        return timetrace, voltages
+        return times, voltages
 
     @rpc_method
-    def get_time_resolution(self, time_base: int) -> int:
-        """Returns the scope's time resolution in Nanoseconds depending on the time_base selector.
+    def get_time_resolution(self, time_base: int) -> float:
+        """Returns the scope's time resolution in nanoseconds depending on the time_base selector.
 
         Parameters:
-            timebase:   Timebase selector (range 2 ... 2**25-1).
-                        The effective time resolution is 4 ns for timebase = 2.
-                        The effective time resolution is (timebase-2)/(125000000) for timebase > 2.
+            time_base:   Timebase selector (range 0 ... 2^25-1).
+                         The effective time resolution is 2^time_base ns for timebase <= 2.
+                         The effective time resolution is (timebase-2)/(125000000) for timebase > 2.
 
         Returns:
-                    Scope's time resolution in Nanoseconds.
-                    0 is returned for invalid time_base.
+            resolution:  Scope's time resolution in nanoseconds.
+                         0 is returned for invalid time_base.
         """
+        if 0 <= time_base <= 2:
+            return 2.0 ** time_base
 
-        if time_base == 2:
-            return 4
-
-        elif 2 < time_base < 2**24:
-            return int(round(((time_base - 2) / 125e6) * 1e9, 0))
+        elif 2 < time_base < 2 ** 24:
```

Comment:
from the API doc I see the time base can go up to 2^32 - 1. Is there a different API doc for this model. I'm using [this](https://www.picotech.com/download/manuals/picoscope-3000-series-a-api-programmers-guide.pdf) one

Reply:
You are right, I'm not sure where that 24 came from. Might have been some hard-coded limit by the experimentalists and it never got changed. I'll change it to 24.